File: hackerrank/problem solving/algorithms/2 - implementation/37 - jumping on the clouds/with morfo library/jumping_on_the_clouds.cs
// https://www.hackerrank.com/challenges/jumping-on-the-clouds/problem?isFullScreen=true

using Morfo.IO;

Reader.SkipInputLn();
int[] clouds = Reader.ReadLn<int, int[]>();
Console.WriteLine(MinJumps(clouds));

// n: length of clouds
// T: O(n)
// S: O(1) extra space
int MinJumps(int[] clouds)
{
    int jumps = 0;
    for (int i = 0, n = clouds.Length - 1; i < n; i += Skip(i, clouds))
        jumps++;
    return jumps;
}

int Skip(int idx, int[] clouds) => IsNextSecondCloudCumulus(idx, clouds) ? 2 : 1;

bool IsNextSecondCloudCumulus(int idx, int[] clouds) => idx + 2 < clouds.Length && clouds[idx + 2] == 0;

File: hackerrank/problem solving/algorithms/2 - implementation/37 - jumping on the clouds/with morfo library/jumping_on_the_clouds.cpp
// https://www.hackerrank.com/challenges/jumping-on-the-clouds/problem?isFullScreen=true
// C++23

import morfo;
import std;
using namespace std;

int min_jumps(const vector<int>& clouds);
int skip(int idx, const vector<int>& clouds);
bool is_next_second_cloud_cumulus(int idx, const vector<int>& clouds);

int main()
{
    morfo::skip_input_ln();
    vector<int> clouds = morfo::readln();
    cout << min_jumps(clouds);
    return 0;
}

// n: length of clouds
// T: O(n)
// S: O(1) extra space
int min_jumps(const vector<int>& clouds)
{
    int jumps{};
    for (int i{}, size = clouds.size() - 1; i < size; i += skip(i, clouds))
        ++jumps;
    return jumps;
}

int skip(int idx, const vector<int>& clouds)
{
    return is_next_second_cloud_cumulus(idx, clouds) ? 2 : 1;
}

bool is_next_second_cloud_cumulus(int idx, const vector<int>& clouds)
{
    return idx + 2 < clouds.size() && clouds.at(idx + 2) == 0;
}
